flow/layers/image_filter_layer_unittests.cc

```diff
@@ -289,12 +289,20 @@ TEST_F(ImageFilterLayerTest, ChildIsCached) {
   EXPECT_EQ(raster_cache()->GetLayerCachedEntriesCount(), (size_t)0);
   EXPECT_FALSE(raster_cache()->Draw(mock_layer.get(), other_canvas));
   EXPECT_FALSE(raster_cache()->Draw(mock_layer.get(), cache_canvas));
+  EXPECT_FALSE(raster_cache()->Draw(layer.get(), other_canvas,
+                                    RasterCacheLayerStrategy::kLayerChildren));
+  EXPECT_FALSE(raster_cache()->Draw(layer.get(), cache_canvas,
+                                    RasterCacheLayerStrategy::kLayerChildren));
 
   layer->Preroll(preroll_context(), initial_transform);
 
   EXPECT_EQ(raster_cache()->GetLayerCachedEntriesCount(), (size_t)1);
   EXPECT_FALSE(raster_cache()->Draw(mock_layer.get(), other_canvas));
   EXPECT_TRUE(raster_cache()->Draw(mock_layer.get(), cache_canvas));
+  EXPECT_FALSE(raster_cache()->Draw(layer.get(), other_canvas,
+                                    RasterCacheLayerStrategy::kLayerChildren));
+  EXPECT_TRUE(raster_cache()->Draw(layer.get(), cache_canvas,
```

Comment:
This seems rather unexpected that both the child layer and the layer's children are both cached. The child layer never asked to be cached, but I guess that is a byproduct of the way the keys are implemented. I don't necessarily have an issue with it, but it was not something I expected to happen.

Reply:
We've lost some common ground here in the discussion.

When I started reviewing this change, I had the following expectation that I am not saying the code was required to match, it's just my preconception when I was reviewing this PR:

Calling Prepare on a layer with kLayerChildren would add that layer to the cache.
(Assume Preroll called here)
Calling Draw on that layer with kLayerChildren would be successful in drawing.
Calling Draw on any of that layer's children would not be successful in drawing because the children themselves are not in the cache, but they are only cached as an aggregate.

Looking at the implementation at the time I made that comment, it looks like those expectations should have been true. In particular, the line in the test that draws the MockLayer should have returned false because even though the entry in the cache had the single MockLayer's ID in it, the type would have been kLayerChildren which would not match the assumed type of the Draw. What I thought your implementation was doing was:

layer->Preroll:
  - adds a cache entry with ({mock_id}, kLayerChildren, cache_transform)

Draw(mock_layer, other_canvas)
  - looks up entry with ({mock_id}, kLayer, other_transform)
  - returns false because that key doesn't match the one in the cache (transform mismatch)
Draw(mock_layer, cache_canvas)
  - looks up entry with ({mock_id}, kLayer, cache_transform)
  - should return false because the key type doesn't match the one in the cache
  - But, you were seeing it pass with EXPECT_TRUE()??????????
  - This was the surprising part, I was expecting that line would probably fail
  - And when looking at the implementation I thought that it was supposed to fail
  - But, it wasn't?
Draw(layer, kLayerChildren, other_canvas)
  - looks up entry with ({mock_id}, kLayerChildren, other_transform)
  - returns false due to transform mismatch
Draw(layer, kLayer, cache_canvas)
  - looks up entry with ({mock_id}, kLayerChildren, cache_transform)
  - This is the only test line that I thought should return true

I was saying that I thought the second test line should probably not have passed and was surprised that it did and wanted to know how it was passing. Instead you rewrote the code so that it likely will pass now, but that wasn't the question and I would have preferred that it did not pass.

Reply:
@flar I think the change you expect is this https://github.com/flutter/engine/pull/31875/commits/edb0adce900f37afa5faa0e358fc479611c130b2. It is done now.

I would like to explain why my previous implementation had this special logic. Imagine this scene, drawing two consecutive frames: frame N and frame N+1.
```
// layer tree of frame N
- TransformLayer
    - ImageFilterLayer (unique_id=51)
        - ColorFilterLayer (unique_id=52)

// layer tree of frame N+1
- TransformLayer
    - ColorFilterLayer (unique_id=52)
```

Before this PR, if `ImageFilterLayer` cached its child `ColorFilterLayer` at frame N, then at frame N+1, `ColorFilterLayer` could hit the cache.  

With this PR, if `ImageFilterLayer` cached its child `ColorFilterLayer` at frame N, then at frame N+1, `ColorFilterLayer` could hit the cache. Because keys `ImageFilterLayer` to cache its own child at frame N and `ColorFilterLayer` to check whether it is cached at frame N+1 are same. ( raster key is  RasterCacheKey(unique_id: 52, type: kLayer))

But in the implementation of this PR, `ColorFilterLayer` cannot hit the cache in frame N+1, Because `ImageFilterLayer` specifies `type` is  `kLayerChildren` when caching.

This is why I added a special logic before. In this scenario, ColorFilterLayer can hit the cache in frame N+1. I'm also not sure if such a scenario would happen in a real flutter app though.

Reply:
I think it would be OK to equate "parent cache my children <single child>" with "<single child> cache me", but we should probably land without that and then pursue that as an optimization later so we can focus on its correctness separately.

Reply:
Got it, that logic is no longer included in the current implementation of this PR. I will file a PR to do it later.